### PR TITLE
Make index attribute of tags writable

### PR DESCRIPTION
**Pull request by t-wissmann**
This requires the respective array operations in the IndexingObject and
also it requires that the EWMH properties are updated accordingly.
While writing test cases for that, I noticed that the EWMH properties
are not updated properly when a tag is removed (this changes the index
of the clients on later tags).

The wiring of the signals now solves both issues.

### CODE REVIEW

File: src/indexingobject.h [hunk 1]
```diff
@@ -83,7 +84,48 @@ class IndexingObject : public Object {
...
+        }
+        int oldIndexSigned = index_of(object);
+        if (oldIndexSigned < 0) {
+            return;
+        }
+        size_t oldIndex = static_cast<size_t>(oldIndexSigned);
+        if (newIndex == oldIndex) {
+            return;
+        }
+        printf("fixing indices\n");
```
Comment (blarz) [inline src/indexingobject.h:106]: Debug leftover?

File: tests/test_ewmh.py [hunk 2]
```diff
@@ -28,6 +28,44 @@ def test_net_wm_desktop_after_bring(hlwm, x11):
...
+def test_net_wm_desktop_after_tag_index_increment(hlwm, x11):
+    hlwm.call('add tag1')
+    hlwm.call('add tag2')
+    hlwm.call('add tag3')
+    # put a new window on the tag with index 1
+    hlwm.call('rule tag=tag1')
+    win, winid = x11.create_client()
+    assert x11.get_property('_NET_WM_DESKTOP', win)[0] == 1
+
+    # move the tag3 to the begin of the list
```
Comment (blarz) [inline tests/test_ewmh.py:40]: beginning

File: tests/test_ewmh.py [hunk 3]
```diff
@@ -28,6 +28,44 @@ def test_net_wm_desktop_after_bring(hlwm, x11):
...
+    assert x11.get_property('_NET_WM_DESKTOP', win)[0] == 3
+
+    if method == 'index_change':
+        # move the tag1 to the end of the list
+        hlwm.call('attr tags.by-name.tag1.index 3')
+    elif method == 'tag_removal':
+        # remove 'tag1' so all later tags experience a index shift
+        hlwm.call('merge_tag tag1 default')
+
+    # so the index of the tag of the client increased by one
```
Comment (blarz) [inline tests/test_ewmh.py:64]: decreased?

**Review by blarz — CHANGES_REQUESTED** [review]

File: src/indexingobject.h (hunk 1 above)
Comment (t-wissmann) [reply src/indexingobject.h:106]: Oh right, thanks!

**Comment by t-wissmann** [comment]
Thanks, I've also fixed the two comments :)

**Comment by blarz** [comment]
While playing around with this change I noticed that default panel.sh plays nicely with this, but I am using xfce4-panel and it seems to be dragging behind somehow:

I am on a tag with just one terminal, in this terminal I change the index of that tag. In the "workspace view" of xfce4-panel and in the list of windows for that workspace I still see that terminal listed under the old index.

Is there maybe something else that needs to get updated or set to "dirty"?

**Comment by t-wissmann** [comment]
Hm there are still multiple bugs, that I can reproduce. I'm on it now.

**Comment by t-wissmann** [comment]
There were two issues:
  - I forgot to update `_NET_CURRENT_DESKTOP` (index of the focused tag)
  - Regarding your issue: Originally, I tried to automatically correct it if the user supplies an index that is too large. However, this implied that I had to do thiese double `index != new_index` checks and in the end I know require that the index supplied by the user is in the interval [0, tag_count)

I've added test cases for both issues

**Review by blarz — APPROVED** [review]
Looks good to me!

**Pull request merged**